### PR TITLE
feat(repo-pulse): keep every merged PR's verification obligation

### PR DESCRIPTION
**Pull request by WingedGuardian**
## Summary

"Run the end-to-end after this merges" lived in whoever merged it. Measured 2026-09-04: one of two owner-directed merges had its E2E forgotten until the owner asked, and **both found something real once run**. #1808 made the decision declarable; the owner then made it advisory, because a block guarantees a sentence *exists*, not that it is *true*. This is the half that closes the hole — the obligation moves after the merge, where nothing has to remember it.

The repo-pulse worker now opens one `pr_verifications` row per merged PR. A documentation-only diff is born closed with its reason recorded: a **deterministic exemption by path**, so no model is ever asked whether a prompt or skill change needs an end-to-end. An unreadable, empty, or API-capped changed-file list **opens** the row — the cost is one validator look, where the other direction silently forgives an unverified merge.

Refs #1718 (half B; half A shipped in #1808).

## Why not `follow_ups`

The reasoning lives in the migration docstring, where the next person will find it. Short version, measured: `status='blocked'` rows reach **ego dispatch** through `get_actionable`, and `'pending'` rows reach the **morning report** through `get_pending`. Every field would have been a small lie, held in check only by twelve readers each remembering a `kind` filter. A narrow table with one writer is a chokepoint; a shared table with per-reader exclusion is a convention — and conventions are what reviewers find one instance of at a time.

## What the review caught

The lane rode the worker's **shared cursor**, so a run that recorded nothing still advanced past those PRs — and the docstring claimed the window would re-cover. It would not: idempotency guarantees a retry cannot duplicate, but nothing retries a PR behind the cursor.

That window is **structural on every existing install**: the run-recording tables predate this one, so between a code deploy and the next server restart `_record_run` succeeds while the lane cannot write. This install would have stranded every merge in that interval, at ~11/day — the exact obligation the feature exists to keep. Verified: the live DB has no `pr_verifications` table today.

The lane now reports whether it recorded its whole window; an incomplete lane fails the run and **keeps the cursor**, the same posture `posted_index_read_failed` already takes for the ledger read.

## Not write-only state

```
python scripts/repo_pulse_worker.py --verification-backlog
```

Plus retention (closed rows only — an open row *is* the obligation, and is never pruned), a `verification_enabled` knob, and the fresh-install schema alongside the migration, pinned equal by a test comparing what SQLite itself recorded.

## Testing

- [x] `ruff check .` passes
- [x] `pytest` — 138 green across the verification lane, crud/schema parity, doc-path parity, and both pre-existing repo-pulse suites
- [x] `docs/architecture/CURRENT.md` updated (entry + `verified:` stamp)

**Against real data, not fixtures.** 105 of 105 real merged PRs recorded, 21 auto-exempted (20% docs-only, consistent with the 23% measured beforehand on a different 60-PR sample); a second run creates 0. Replaying the deploy window: pre-migration reports incomplete so the cursor holds → migration lands → **all 105 stranded PRs recovered** → re-run creates 0.

**Ground truth, both directions.** PR #1812 (`.claude/skills/**/SKILL.md`, a prompt surface) auto-exempted — precisely the case that should be decided by rule rather than by a model. PR #1808 (code + docs) correctly left open.

**19 of 19 mutations RED**, zero survivors, hash-verified restores. Four of them mutate the *review fix itself* — which is how a missing test for the empty-reason guard was caught: it survived the first sweep, so the guard was unlocked.

## Known and deliberate

The **consumer does not exist yet**. The Wave-3 validator session is what closes open rows; until it lands the backlog grows and `--verification-backlog` is the only reader. That was decided with the volume measured in advance, and nothing here claims otherwise.

On an existing install, recording starts from merges the worker sees **after** the upgrade — the lane rides the shared cursor, so it does not reach back over history already behind it. A full-window backfill would need a deliberate cursor-clearing run, which is a data repair and is not performed here.

E2E: run `python scripts/repo_pulse_worker.py --trigger manual --force` after the next server restart, then `--verification-backlog`; expect rows for merges since the cursor, docs-only ones already closed.

## Summary by CodeRabbit

- **New Features**
  - Added durable post-merge pull request verification tracking.
  - Documentation-only changes can be automatically marked complete; other changes remain available for validation.
  - Added independent controls for enabling verification and a command to review outstanding items.
  - Verification progress now recovers safely when switching repositories or encountering incomplete results.

- **Bug Fixes**
  - Improved backlog reporting with repository names, truncation notices, and safe handling when no database is available.
  - Prevented invalid retention periods from starting cleanup operations.
  - Improved handling of incomplete or malformed changed-file data.

### CODE REVIEW

**Comment by WingedGuardian** [comment]
Structural review and PII/Secrets/Wording scan for PR #1836 (head `df827cf07e375e9bb409f07d488c79139fc44d04`).

### Structural Review

No blocking pattern violations found across the 17 changed files. Specifically checked and clear:
- No `nohup`, no direct inserts into `task_states`, no `/tmp/` temp-file usage, no `git add .`/`-A`, no `git push --force`/`reset --hard`, no `pip install -e` to a worktree path, no bare `except:` clauses, no mutable default arguments, no unguarded `os.killpg`, and no raw SQL against `genesis.db` outside the CRUD/migration boundary (the new `src/genesis/db/crud/pr_verifications.py` is itself the CRUD module, so its SQL is the intended location; `scripts/repo_pulse_worker.py`'s new `--verification-backlog` reader goes through `verif_crud.list_open`/`counts`, not raw SQL).
- Import hygiene is clean: no circular imports, no star imports, stdlib/third-party/local grouping correct in `pr_verifications.py`, `doc_paths.py`, and the new imports added to `repo_pulse_worker.py`.
- New modules (`src/genesis/session_awareness/doc_paths.py`, `src/genesis/db/crud/pr_verifications.py`, the `20260906234824_pr_verifications` migration) all have matching test files (`test_doc_paths.py`, `test_pr_verifications.py`, `test_repo_pulse_verification_lane.py`), including a fresh-install-vs-migration schema parity test and a hook-vs-lane classifier parity test.
- Public functions/classes are documented; docstrings explain rationale, not just mechanics.
- PR title carries a conventional commit prefix (`feat(repo-pulse): ...`).

One non-blocking coverage NOTE:
- `scripts/prune_repo_pulse.py`'s `_prune()` wrapper changed shape (now threads `verification_days`, returns a 2-tuple, and changed its print format) but has no dedicated test file. The underlying logic it calls (`pr_verifications.prune_closed`) is well covered in `test_pr_verifications.py::test_prune_deletes_only_old_closed_rows_never_open_ones`, so this is only a gap in the thin CLI wrapper — and it's a pre-existing gap (the script had no tests before this PR either), not a regression.

### PII/Secrets/Wording

Scanned all added lines plus PR title/body/CHANGELOG:
- No API keys, tokens, passwords, secrets, private IPs (10.x/192.168.x/100.64–127.x, fd00::/8), SSH keys, or credentialed connection strings.
- No personal emails, usernamed absolute paths, location-revealing timezone identifiers, cloud account/project IDs, hardware identifiers, or private hostnames.
- PR title, body, and the `CHANGELOG.md` addition describe the feature (durable post-merge verification tracking) in generic terms — no mention of a PII/secrets leak, security incident, named private infrastructure, or exploitable vulnerability detail.
- PII/Secrets/Wording: clean, no soft warnings either.

VERDICT: PASS



---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/prune_repo_pulse.py [hunk 1]
```diff
@@ -40,10 +50,23 @@ def main() -> None:
         default=45,
         help="retention window in days (rows older than this are deleted)",
     )
+    ap.add_argument(
+        "--verification-days",
+        type=int,
+        default=180,
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/prune_repo_pulse.py:78]: Fixed, and in two places rather than one. `prune_closed` now raises on `older_than_days < 1` — the guard belongs in the CRUD because the caller that gets this wrong is the one that never thought about it, and it mirrors `prune_merge_journal` (crud/entities.py:550), which names this exact class for the entity journal. The CLI additionally takes a validated argparse type so the operator gets a readable refusal during parsing, before any connection opens.

Widened beyond the reported flag: `--days` had the identical defect and was never reported. Both take the type — they share the arithmetic, so they shared the bug.

Verify-RED: removing the CRUD guard turns 3 of 3 negative-window cases red, while the `days=1` boundary case correctly stays green. Reverting the flags to `type=int` turns the 2 subprocess cases red while the in-process ones stay green — which is why the subprocess test exists.

File: src/genesis/session_awareness/repo_pulse_worker.py [hunk 2]
```diff
@@ -959,6 +1052,61 @@ def _base_row(**over) -> dict:
...
+    # timeout, and this lane must run on every tick that has new PRs. Own
+    # try/except + knob (the open-PR-lane posture): a gh/DB failure surfaces on
+    # the run row's detail and never breaks the absorb lanes or _record_run.
+    # Rides its own dedup (the (repo, pr_number) unique index), so the cursor
+    # advancing past a failed tick is recovered by the next window re-coverage
+    # only when the failure was per-PR transient — a hard lane failure is
+    # visible in detail and the rows self-heal on the next tick that sees the
+    # same PRs; PRs the cursor has passed are backfilled by nothing, which is
+    # why the failure note matters.
+    if cfg.get("verification_enabled", True):
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/session_awareness/repo_pulse_worker.py]: Fixed as a class — but the reasoning in the finding is wrong, and the correction matters.

The claim is that this knob is not type-validated "like its sibling knobs". It isn't inconsistent with them: the twin `open_pr_enabled` uses the identical bare `cfg.get(..., True)` at repo_pulse_worker.py:605, and the typed helpers (`knob_int`, `knob_float01`) exist for int/float knobs, not bools. All three bool knobs shared the defect.

The concern underneath is real. `cfg.get(key, True)` is a truthiness test, and the values an operator actually mistypes are truthy STRINGS — a quoted "false", "no" or "off" all read as ENABLED, so the switch does the opposite of what was asked and says nothing. (Unquoted `off` is the lucky case: YAML 1.1 gives a real False.)

New `knob_bool`, applied to all three sites — `verification_enabled`, `open_pr_enabled`, and the master `enabled`. Fixing one of three would have recreated the same bug in the same file.

It degrades to the DEFAULT and logs, rather than to off: this module's stated rule for an invalid `mode` is "never a silent off", and an unrecorded obligation is worse than a noisy one. The defect being fixed is the silence, not the direction.

Verify-RED: reverting to bare truthiness turns all 5 truthy-string cases red.

File: scripts/repo_pulse_worker.py [hunk 3]
```diff
@@ -30,6 +30,44 @@
...
+
+    from genesis.db.crud import pr_verifications as verif_crud
+    from genesis.env import genesis_db_path
+
+    resolved = db_path or str(genesis_db_path())
+
+    async def _read() -> tuple[list[dict], dict]:
+        import aiosqlite
+
+        async with aiosqlite.connect(resolved, timeout=10) as db:
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/repo_pulse_worker.py]: Fixed. A report-only command held a read-WRITE handle, so pointing it at a wrong path CREATED an empty database and then truthfully reported it as empty. Now `file:...?mode=ro` with `uri=True`.

`mode=ro` deliberately, not `immutable=1`: immutable ignores the -wal file, so it would miss rows the worker committed moments earlier — a read-only report that silently under-reports is worse than the write handle was. A missing file is now reported as "no database at <path>" rather than raising, so "not set up yet" doesn't read as a defect.

File: scripts/repo_pulse_worker.py [hunk 4]
```diff
@@ -30,6 +30,44 @@
...
+            await db.execute("PRAGMA busy_timeout=5000")
+            db.row_factory = aiosqlite.Row
+            return await verif_crud.list_open(db), await verif_crud.counts(db)
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/repo_pulse_worker.py:76]: Fixed, and it's the repo's own "select, don't amputate" rule rather than a nitpick: `list_open` caps at 500 while the histogram beneath prints the TRUE open total, so at 501 open rows the command printed 500 lines and said 501 — two numbers disagreeing in silence, where counting the lines gives a wrong answer that looks complete.

The cut is now declared with both numbers, which are known exactly here (the histogram is a separate COUNT, so this is the "total is known" case, not the "state the cap and mark the tail unknown" case). Kept the cap rather than paging: an unbounded print of an unbounded store is the defect one level over.

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
@codex review

File: config/repo_pulse.yaml [hunk 5]
```diff
@@ -71,6 +71,18 @@ inject_confidence_floor: 0.7
...
+# One durable pr_verifications row per MERGED PR: the post-merge E2E obligation
+# that survives the merge (the merge-time E2E: declaration is advisory). A
+# documentation-only diff is auto-closed with the reason recorded — a
+# DETERMINISTIC exemption by path (doc_paths.is_doc_path), never a model call.
+# Everything else stays open for the validator session. An unreadable
+# changed-file list fails toward KEEPING the obligation (the row opens).
+# Same lever ladder as open_pr_enabled: this knob stops ONLY this lane; the
+# enabled/mode full-worker stops apply on top. Backlog reader:
+#   python scripts/repo_pulse_worker.py --verification-backlog
+verification_enabled: true
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply config/repo_pulse.yaml:84]: Verified and fixed — as the population rather than the instance. The validator now derives its boolean keys from a single `_BOOL_KNOBS` registry in `repo_pulse_config`, mirroring what `_INT_KNOBS` already did, plus a coverage test that fails if a boolean lands in DEFAULTS unregistered. Adding just the one name would have left the next knob to repeat it.

File: src/genesis/session_awareness/repo_pulse_worker.py [hunk 6]
```diff
@@ -959,6 +1053,61 @@ def _base_row(**over) -> dict:
...
+    # timeout, and this lane must run on every tick that has new PRs. Own
+    # try/except + knob (the open-PR-lane posture): a gh/DB failure surfaces on
+    # the run row's detail and never breaks the absorb lanes or _record_run.
+    # Rides its own dedup (the (repo, pr_number) unique index), so the cursor
+    # advancing past a failed tick is recovered by the next window re-coverage
+    # only when the failure was per-PR transient — a hard lane failure is
+    # visible in detail and the rows self-heal on the next tick that sees the
+    # same PRs; PRs the cursor has passed are backfilled by nothing, which is
+    # why the failure note matters.
+    if knob_bool(cfg, "verification_enabled"):
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/session_awareness/repo_pulse_worker.py]: Verified and fixed. The lane now keeps its own `verification_through` watermark that advances only on a tick where it ran to completion, reads its window from that watermark, and widens the fetch to reach back to it. An absent watermark is deliberately NOT read as caught-up: the first tick after deploy re-covers the ordinary lookback so PRs stranded by an earlier disabled window are recovered — dedup-safe via the unique index, and bounded by lookback_days, which is the stated limit. My own acceptance test then caught the fix being inert: the no-new-PRs early return fires before the lane, i.e. on exactly the recovery ticks, so it now tests the lane window too.

File: scripts/repo_pulse_worker.py [hunk 7]
```diff
@@ -30,6 +30,64 @@
...
+
+    async def _read() -> tuple[list[dict], dict]:
+        import aiosqlite
+
+        # mode=ro, not a plain path: this command only REPORTS. A read-write
+        # handle would create an empty database when pointed at a wrong path and
+        # then truthfully report it as empty. `mode=ro` (not `immutable=1`) is
+        # the WAL-aware read-only form — `immutable` ignores the -wal file and
+        # would miss rows the worker committed moments earlier.
+        async with aiosqlite.connect(f"file:{resolved}?mode=ro", uri=True, timeout=10) as db:
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/repo_pulse_worker.py]: Verified and fixed. The URI is built with `Path.as_uri()` and the parameter appended, so a path containing `?` or `#` can no longer be parsed as a query or fragment — which, as you noted, could also swallow `mode=ro` and turn a report-only read into a handle that creates the unintended file.

File: scripts/repo_pulse_worker.py [hunk 8]
```diff
@@ -30,6 +30,64 @@
...
+            db.row_factory = aiosqlite.Row
+            return await verif_crud.list_open(db), await verif_crud.counts(db)
+
+    rows, histogram = _asyncio.run(_read())
+    if not histogram:
+        print("pr_verifications: no rows (table empty or pre-migration)")
+        return
+    for row in rows:
+        title = (row.get("pr_title") or "").strip()
+        print(f"OPEN  PR #{row['pr_number']}  merged {str(row['merged_at'])[:10]}  {title[:80]}")
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/repo_pulse_worker.py]: Verified and fixed. Each backlog line now prints `repo#number`, so the row is identified by its full (repo, pr_number) key and the operator has the `repo` argument `close_verification` requires.

File: src/genesis/session_awareness/repo_pulse_worker.py [hunk 9]
```diff
@@ -550,7 +679,18 @@ def _base_row(**over) -> dict:
...
+    # verification lane could never see a PR the shared cursor already passed
+    # — the re-coverage above would be a window that never contains anything.
+    # When the lane is disabled its watermark is irrelevant: nothing will read
+    # those PRs, so paying for a wider gh query would be waste.
+    verification_on = knob_bool(cfg, "verification_enabled")
+    fetch_from = cursor_before
+    if verification_on and (verification_before is None or not cursor_before):
+        fetch_from = None  # fall back to the plain lookback (bounded)
+    elif verification_on and verification_before:
+        fetch_from = min(str(cursor_before), str(verification_before))
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/session_awareness/repo_pulse_worker.py:763]: Addressed as the CLASS at `f14721b5b`. All five of this round share one cause: the previous fix added a second window to machinery built for one, and every mechanism that verifies completeness was written for the shared cursor. The invariant is now stated once — the lane advances only over a window it can prove complete. For this finding specifically: the capped-window path returns above the advance, failing the run and holding both cursors, which is now stated in the code rather than left implicit since a deep backlog is exactly the path that takes.

File: src/genesis/session_awareness/repo_pulse_worker.py [hunk 10]
```diff
@@ -120,16 +135,30 @@ def _write_cursor(root: Path, prior: dict, *, merged_at: str | None) -> None:
...
+    never opened. It advances ONLY on a tick where the lane actually ran to
+    completion, so a disabled (or failed) lane leaves it standing while
+    ``last_merged_at`` moves on.
     """
     last = prior.get("last_merged_at")
     if merged_at is not None:
         last = max(str(last), merged_at) if last else merged_at
+    verif = prior.get("verification_through")
+    if verification_through is not None:
+        verif = max(str(verif), verification_through) if verif else verification_through
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/session_awareness/repo_pulse_worker.py]: Fixed at `f14721b5b`. You are right that identity is (repo, pr_number) while the watermark was a bare timestamp. The repo is stored WITH the watermark now, and a mismatch DISCARDS it rather than applying it — re-covering the lookback is free because the unique index absorbs it. Pinned by a test that puts another repository's watermark in the cursor and asserts an older PR here still gets its row.

File: src/genesis/db/crud/pr_verifications.py [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+# Today's entry points touch one DB per process, so this is prophylactic; it
+# costs one dict lookup and removes a trap rather than documenting it.
+_tables_verified: set[str] = set()
+
+
+def _db_key(db: aiosqlite.Connection) -> str:
+    # aiosqlite exposes the path it was constructed with; fall back to the
+    # object id so an unexpected shape degrades to "cache nothing" rather than
+    # aliasing two databases onto one key.
+    return str(getattr(db, "_conn_path", None) or getattr(db, "_path", None) or id(db))
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/db/crud/pr_verifications.py]: Fixed at `f14721b5b`, and the comment was worse than the code. It claimed the `id(db)` fallback "degrades to cache nothing" — it does the opposite, since object ids are good dict keys and get reused after close, so a later connection to a different database could be told the tables exist without consulting `sqlite_master`. `_db_key` returns None when the path is unknown, and None now genuinely means do not cache.

File: src/genesis/session_awareness/repo_pulse_worker.py [hunk 12]
```diff
@@ -609,13 +749,33 @@ def _base_row(**over) -> dict:
...
+    if verification_on:
+        verif_prs = sorted(
+            (
+                p
+                for p in prs_by_number.values()
+                if not verification_before or str(p["mergedAt"]) > str(verification_before)
+            ),
+            key=lambda p: str(p["mergedAt"]),
+        )
+    if not prs and not verif_prs:
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/session_awareness/repo_pulse_worker.py:870]: Fixed at `f14721b5b`. Correct — when the shared cursor is current but the watermark is behind, `prs` is empty while `verif_prs` is not, and the fuzzy tier would call the judge with no PRs at all: pure cost, and a timeout fails the run before the watermark is written, so the window just recorded repeats every tick. A lane-only tick now skips the tiers that have nothing to match.

File: src/genesis/session_awareness/repo_pulse_worker.py [hunk 13]
```diff
@@ -609,13 +749,33 @@ def _base_row(**over) -> dict:
...
+    # not: returning early on `not prs` would skip the lane on exactly the
+    # ticks it exists to catch up on, leaving the fix inert.
+    verif_prs: list[dict] = []
+    if verification_on:
+        verif_prs = sorted(
+            (
+                p
+                for p in prs_by_number.values()
+                if not verification_before or str(p["mergedAt"]) > str(verification_before)
+            ),
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/session_awareness/repo_pulse_worker.py:867]: Fixed at `f14721b5b` — this is the finding that named the generator. `list_merged_prs` now reports what its row-validation filter discarded, and any discard makes the window incomplete, so the watermark cannot advance past a merge the listing silently lost. VERIFY-RED: removing that conjunct fails the new test while its control stays green, which is what distinguishes the fix from a watermark that simply never advances.

File: src/genesis/session_awareness/repo_pulse_worker.py [hunk 14]
```diff
@@ -120,16 +137,38 @@ def _write_cursor(root: Path, prior: dict, *, merged_at: str | None) -> None:
...
+    never opened. It advances ONLY on a tick where the lane actually ran to
+    completion, so a disabled (or failed) lane leaves it standing while
+    ``last_merged_at`` moves on.
     """
     last = prior.get("last_merged_at")
     if merged_at is not None:
         last = max(str(last), merged_at) if last else merged_at
+    verif = prior.get("verification_through")
+    if verification_through is not None:
+        verif = max(str(verif), verification_through) if verif else verification_through
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/session_awareness/repo_pulse_worker.py]: Confirmed, and fixed on both sides of the pair rather than the one that was flagged.

The reader's reset was already there; the writer stayed monotonic, so a foreign watermark newer than the new repository's newest merge survived the max and was then stamped with the current repository's name — after which the reader's guard could never fire again. That is the finding exactly.

`_write_cursor` now takes an explicit reset flag. On a repository change it clears the watermark outright instead of advancing or preserving it, so the next tick derives its window from the plain lookback and re-covers. The reset tick deliberately does not advance the watermark either: the window it fetched came from the old repository's cursor and may be narrower than the lookback. Outside a reset, the monotonic max is now guarded by an identity comparison, so it only ever applies within one repository.

On the test: a companion case asserts the whole contract rather than half of it — the cursor is cleared on the reset tick and records the current repository, and it advances again on the next tick. Asserting only the clear would pass for a watermark that never recovers at all.

One correction to the proposed diff: deriving the reset from a comparison inside the writer alone leaves the reader and writer inferring the same fact independently, which is what produced this defect. The flag is computed once, where identity is already known, and passed down.
